Add support for WASI reactor in pure Zig-exe.

mathetake opened the pull request:
This PR adds support for WASI reactor in pure Zig WASI targets. Notably this enables users to produce WASI reactor binary with `build-exe` command with `-mexec-model=reactor` just like `zig cc`, `zig c++`, and clang.

With `-mexec-model=reactor` flag, Zig 1) exports `_initialize` symbol which calls the user defined `pub fn main`, 2) does not define `_start`, and 3) does not execute `proc_exit` after `main`.

Please refer to https://github.com/WebAssembly/WASI/blob/main/design/application-abi.md for the ABI in detail.

Finally resolves #6757 together with #9052.

cc @kubkon

kubkon commented on src/main.zig:
```diff
@@ -616,7 +617,7 @@ fn buildOutputType(
     var major_subsystem_version: ?u32 = null;
     var minor_subsystem_version: ?u32 = null;
-    var wasi_exec_model: ?wasi_libc.CRTFile = null;
+    var wasi_exec_model: ?[]const u8 = null;
```
Hmm, I'd rather avoid passing strings instead of enums. How about we create an intermediate enum like you did before which then we can map to `wasi_libc.CRTFile` where required, and we can switch on in the linker rather than compare strings?

mathetake replied:
yeah that's what actually I was thinking about first. Thanks for the suggestion! will fix accordingly.

mathetake replied:
PTAL at https://github.com/ziglang/zig/pull/9178/commits/6769659dedde8b53999fda15b0125639cc451c23

kubkon commented on src/wasi_libc.zig:
```diff
@@ -45,8 +45,20 @@ pub fn emulatedLibCRFileLibName(crt_file: CRTFile) []const u8 {
 
-pub fn crtFileFullName(crt_file: CRTFile) []const u8 {
-    return switch (crt_file) {
+pub fn getExecModelCRTFIle(wasi_exec_model: ?[]const u8) CRTFile {
```
See my comment above: I really would avoid strings at all cost especially since WASI exec model has well defined variants.

mathetake replied:
I believe this is fixed! https://github.com/ziglang/zig/pull/9178/commits/6769659dedde8b53999fda15b0125639cc451c23 addressed this comment.

kubkon commented on lib/std/start.zig:
```diff
@@ -16,7 +16,7 @@ const native_os = builtin.os.tag;
 var argc_argv_ptr: [*]usize = undefined;
 
-const start_sym_name = if (native_arch.isMIPS()) "__start" else "_start";
+const start_sym_name = if (native_arch.isMIPS()) "__start" else if (builtin.is_wasi_reactor_exec_model) "_initialize" else "_start";
```
Hmm, here's a thought. Maybe we should augment the builtins and put the exec model enum there much like this is the case with `GlobalLinkage`, etc.?

mathetake replied:
Thanks! I believe https://github.com/ziglang/zig/pull/9178/commits/6769659dedde8b53999fda15b0125639cc451c23 addressed this comment.

kubkon reviewed (changes requested):
Thanks @mathetake! I've got a few comments I'd like to see addressed before going forward with this PR.

mathetake commented on src/Compilation.zig:
```diff
@@ -3638,6 +3638,15 @@ pub fn generateBuiltinZigSource(comp: *Compilation, allocator: *Allocator) Alloc
         std.zig.fmtId(@tagName(comp.bin_file.options.machine_code_model)),
     });
 
+    const wasi_exec_model_fmt = if (comp.bin_file.options.wasi_exec_model) |model|
+        std.zig.fmtId(@tagName(model))
+    else
+        std.zig.fmtId(@tagName(std.builtin.WasiExecModel.command));
+    try buffer.writer().print(
+        \\pub const wasi_exec_model = std.builtin.WasiExecModel.{};
+        \\
+    , .{wasi_exec_model_fmt});
```
FYI: If I put this into a single statement, 

```zig
    try buffer.writer().print(
        \\pub const wasi_exec_model = std.builtin.WasiExecModel.{};
        \\
    , .{if (comp.bin_file.options.wasi_exec_model) |model|
        std.zig.fmtId(@tagName(model))
    else
        std.zig.fmtId(@tagName(std.builtin.WasiExecModel.command))});
```

I got some weird error around LLVM: 
```
make install
[ 77%] Built target embedded_softfloat
[ 78%] Built target opt_c_util
[ 83%] Built target zigcpp
[ 97%] Built target zigstage1
[ 98%] Built target zig0
[ 99%] Building self-hosted component /home/mathetake/zig/build/zig1.o
broken LLVM module found: Instruction does not dominate all uses!
  %369 = getelementptr inbounds %"Compilation.struct:3644:8", %"Compilation.struct:3644:8"* %39, i32 0, i32 0, !dbg !75434
  call fastcc void @std.zig.fmt.fmtId(%"std.fmt.Formatter(std.zig.fmt.formatId)"* sret(%"std.fmt.Formatter(std.zig.fmt.formatId)") %369, %"[]u8"* @13528), !dbg !75435
Instruction does not dominate all uses!
  %369 = getelementptr inbounds %"Compilation.struct:3644:8", %"Compilation.struct:3644:8"* %39, i32 0, i32 0, !dbg !75434
  %372 = bitcast %"std.fmt.Formatter(std.zig.fmt.formatId)"* %369 to i8*, !dbg !75435
Instruction does not dominate all uses!
  %369 = getelementptr inbounds %"Compilation.struct:3644:8", %"Compilation.struct:3644:8"* %39, i32 0, i32 0, !dbg !75434
  %373 = bitcast %"std.fmt.Formatter(std.zig.fmt.formatId)"* %369 to i8*, !dbg !75435

This is a bug in the Zig compiler.
Aborted (core dumped)
make[2]: *** [CMakeFiles/zig.dir/build.make:328: zig1.o] Error 134
make[1]: *** [CMakeFiles/Makefile2:148: CMakeFiles/zig.dir/all] Error 2
make: *** [Makefile:130: all] Error 2
```

kubkon replied:
I believe this is because it's not enough to just add an enum to `std.builtin`. You need to make more tweaks to stage1 in C++. For starters, we need to have the exact same enum shared between both stage1 and stage2 which means you need add a corresponding enum to `src/stage1/stage1.h`. Then, you'll need to tweak logic for actually marshalling the data between stage1 and stage2 - have a look how this is for `CodeModel` or some other builtin structs.

If you get stuck, I'll be happy to help you out but will probably take a look at this in more detail tomorrow. Alternatively, we can also get @andrewrk involved if he's got a minute.

mathetake replied:
Ah I see. Thanks for clarifying!

mathetake replied:
so yeah I found that this also applies to other existing builtin types: 

```
const std = @import("std");

pub fn main() void {
    std.debug.print("test {}", .{if (@rem(std.time.timestamp(), 2) == 0)
        std.zig.fmtId(@tagName(std.builtin.CodeModel.tiny))
    else
        std.zig.fmtId(@tagName(std.builtin.CodeModel.default))});
}
```

HEAD and 0.8.0 fails to compile this code with the same error like this:

```
zig build-exe main.zig
broken LLVM module found: Instruction does not dominate all uses!d.debug.ModuleDebug...
  %2 = getelementptr inbounds %"struct:4:33", %"struct:4:33"* %0, i32 0, i32 0, !dbg !16731
  call fastcc void @std.zig.fmt.fmtId(%"std.fmt.Formatter(std.zig.fmt.formatId)"* sret(%"std.fmt.Formatter(std.zig.fmt.formatId)") %2, %"[]u8.18"* @390), !dbg !16733
Instruction does not dominate all uses!
  %2 = getelementptr inbounds %"struct:4:33", %"struct:4:33"* %0, i32 0, i32 0, !dbg !16731
  %5 = bitcast %"std.fmt.Formatter(std.zig.fmt.formatId)"* %2 to i8*, !dbg !16733
Instruction does not dominate all uses!
  %2 = getelementptr inbounds %"struct:4:33", %"struct:4:33"* %0, i32 0, i32 0, !dbg !16731
  %6 = bitcast %"std.fmt.Formatter(std.zig.fmt.formatId)"* %2 to i8*, !dbg !16733

This is a bug in the Zig compiler.thread 3208559 panic:
Unable to dump stack trace: debug info stripped
zsh: abort (core dumped)  zig build-exe main.zig
```

and the same, if we split the line to two statements. the error disappears.

mathetake replied:
fyi: I tweaked the C++ and around stage1.zig and Compilation.zig following `CodeModel`, but the error has never gone.

mathetake replied:
Luckily the code with split statements just works and successfully produced Wasi Reactors!

mathetake commented:
Thanks @kubkon for a quick look:) I hope https://github.com/ziglang/zig/pull/9178/commits/6769659dedde8b53999fda15b0125639cc451c23 addressed your comments! PTAL!

kubkon reviewed (approved):
Looks great to me, thanks @mathetake! I just need to double check with @andrewrk that we don't need to add matching logic to the C++ codebase for setting `WasiExecModel` builtin.

mathetake commented:
@andrewrk @kubkon I would appreciate it if you have any chance to take a look and upstream this! Thanks a lot:)

andrewrk commented:
Ah I'm sorry for taking so long to get to this! Looking now.

andrewrk commented on lib/std/start.zig:
```diff
@@ -16,7 +16,7 @@ const native_os = builtin.os.tag;
 var argc_argv_ptr: [*]usize = undefined;
 
-const start_sym_name = if (native_arch.isMIPS()) "__start" else "_start";
+const start_sym_name = if (native_arch.isMIPS()) "__start" else if (builtin.wasi_exec_model == .reactor) "_initialize" else "_start";
```
`builtin.wasi_exec_model` should only be accessed if the target OS is WASI. Perhaps this line should be left alone and logic should be added below on line 68?

andrewrk commented on src/wasi_libc.zig:
```diff
@@ -45,8 +45,18 @@ pub fn emulatedLibCRFileLibName(crt_file: CRTFile) []const u8 {
+        CRTFile.crt1_o;
+}
+
+pub fn execModelCRTFileFullName(wasi_exec_model: ?std.builtin.WasiExecModel) []const u8 {
```
```suggestion
pub fn execModelCrtFileFullName(wasi_exec_model: ?std.builtin.WasiExecModel) []const u8 {
```

andrewrk commented on src/wasi_libc.zig:
```diff
@@ -45,8 +45,18 @@ pub fn emulatedLibCRFileLibName(crt_file: CRTFile) []const u8 {
 
-pub fn crtFileFullName(crt_file: CRTFile) []const u8 {
-    return switch (crt_file) {
+pub fn getExecModelCRTFIle(wasi_exec_model: ?std.builtin.WasiExecModel) CRTFile {
```
```suggestion
pub fn execModelCrtFile(wasi_exec_model: ?std.builtin.WasiExecModel) CRTFile {
```

my mistake for calling it CRTFile instead of CrtFile. I will fix that later.

andrewrk commented on src/Compilation.zig:
```diff
@@ -3638,6 +3638,15 @@ pub fn generateBuiltinZigSource(comp: *Compilation, allocator: *Allocator) Alloc
+    else
+        std.zig.fmtId(@tagName(std.builtin.WasiExecModel.command));
+    try buffer.writer().print(
+        \\pub const wasi_exec_model = std.builtin.WasiExecModel.{};
```
Rather than populating the value arbitrarily with "command" when the OS is not WASI, this declaration should be entirely omitted when the OS is not WASI.

andrewrk commented on src/link.zig:
```diff
@@ -119,7 +119,7 @@ pub const Options = struct {
 
     /// WASI-only. Type of WASI execution model ("command" or "reactor").
-    wasi_exec_model: ?wasi_libc.CRTFile = null,
+    wasi_exec_model: ?std.builtin.WasiExecModel,
```
Rather than having this be optional, I suggest that it should be non-optional, and set to `undefined` when the target OS is not WASI.

andrewrk commented on src/main.zig:
```diff
@@ -1065,6 +1066,13 @@ fn buildOutputType(
+                            wasi_exec_model = .reactor;
+                        } else if (std.mem.eql(u8, value, "command")) {
+                            wasi_exec_model = .command;
+                        }
```
Missing an error when a value other than "reactor" or "command" is specified. I suggest using `std.meta.stringToEnum` (look for other usages of that function nearby).

andrewrk commented on src/main.zig:
```diff
@@ -1272,9 +1280,9 @@ fn buildOutputType(
                         } else if (std.mem.eql(u8, it.only_arg, "command")) {
-                            wasi_exec_model = .crt1_command_o;
+                            wasi_exec_model = .command;
                         }
```
Same here, missing an error for bad exec model name.

andrewrk commented on src/wasi_libc.zig:
```diff
@@ -45,8 +45,18 @@ pub fn emulatedLibCRFileLibName(crt_file: CRTFile) []const u8 {
+            .command => CRTFile.crt1_command_o,
+        }
+    else
+        CRTFile.crt1_o;
```
Based on my research it looks like every WASI program is either a command or a reactor. Is it meaningful for a program to be neither? Why is crt1_o even an option?

mathetake replied:
yeah we should remove crt1_o in favor of crt1_command_o, where the latter lets wasm-ld call ctors rather than explicitly calling in crt.c. Also I believe this reduces the complexity here about null vs default stuff.

so I removed the support for crt1_o. cc @kubkon wdyt?

andrewrk commented on src/Compilation.zig:
```diff
@@ -724,7 +724,7 @@ pub const InitOptions = struct {
     test_name_prefix: ?[]const u8 = null,
     subsystem: ?std.Target.SubSystem = null,
     /// WASI-only. Type of WASI execution model ("command" or "reactor").
-    wasi_exec_model: ?wasi_libc.CRTFile = null,
```
Inside the create function of Compilation, it should resolve this `?WasiExecModel` into a `WasiExecModel` - that is - it should choose a default, and then the link Options struct should store only the non-optional WasiExecModel.

andrewrk reviewed (changes requested):
Thanks for the contribution!

I have some requested improvements.

mathetake commented:
Thank you so much @andrewrk, I hope I addressed your suggestion in https://github.com/ziglang/zig/pull/9178/commits/85b5503e3a3f39a37eeb4df790cfe7370433f1de, and now it looks much better:)

PTAL when you have time, thanks!

andrewrk commented on lib/std/start.zig:
```diff
@@ -64,8 +64,9 @@ comptime {
-                if (!@hasDecl(root, start_sym_name)) @export(wasm_freestanding_start, .{ .name = start_sym_name });
+            } else if (native_arch.isWasm()) {
+                const wasm_start_sym = if (builtin.wasi_exec_model == .reactor) "_initialize" else "_start";
+                if (!@hasDecl(root, start_sym_name)) @export(wasm_start, .{ .name = wasm_start_sym });
```
```suggestion
                if (!@hasDecl(root, wasm_start_sym)) @export(wasm_start, .{ .name = wasm_start_sym });
```

andrewrk commented on src/Compilation.zig:
```diff
@@ -3638,6 +3641,14 @@ pub fn generateBuiltinZigSource(comp: *Compilation, allocator: *Allocator) Alloc
         std.zig.fmtId(@tagName(comp.bin_file.options.machine_code_model)),
     });
 
+    if (comp.getTarget().os.tag == .wasi) {
```
```suggestion
    if (target.os.tag == .wasi) {
```

andrewrk commented on src/wasi_libc.zig:
```diff
@@ -45,9 +45,15 @@ pub fn emulatedLibCRFileLibName(crt_file: CRTFile) []const u8 {
-pub fn crtFileFullName(crt_file: CRTFile) []const u8 {
-    return switch (crt_file) {
-        .crt1_o => "crt1.o",
+pub fn getExecModelCrtFIle(wasi_exec_model: std.builtin.WasiExecModel) CRTFile {
```
```suggestion
pub fn execModelCrtFile(wasi_exec_model: std.builtin.WasiExecModel) CRTFile {
```

andrewrk reviewed (changes requested):
Thanks for making the requested changes - just a couple small things left and then this looks ready to merge to me :)

mathetake commented:
thanks for the quick review! addressed https://github.com/ziglang/zig/pull/9178/commits/943f6568a8c8e56b661b1db396d8b8ac89d3e9c3 here :)

andrewrk reviewed (approved):
Alright, this is ready to be merged, did you want to wait until @kubkon wakes up to discuss the question you asked? If not I'm happy to merge now.

mathetake commented:
I can discuss offline with @kubkon later, so I'm happy to merge now. I believe it's easy to bring crt1.o support even if we want later:)
